File: instrumentation/nodejs/instrument.js
const { execSync } = require("child_process");

function installDeps() {
  execSync("npm install --save @opentelemetry/api @opentelemetry/sdk-node @opentelemetry/auto-instrumentations-node", { stdio: "inherit" });
}

function wrapApp(entry) {
  const wrapper = `
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const sdk = new NodeSDK({
  traceExporter: new (require('@opentelemetry/exporter-trace-otlp-grpc').OTLPTraceExporter)({
    url: 'http://otel-collector:4317',
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});
sdk.start()
  .then(() => require('./${entry}'))
  .catch(err => console.error(err));
`;

  require("fs").writeFileSync("otel-wrapper.js", wrapper);
  console.log("Instrumentation added. Run using: node otel-wrapper.js");
}

const entry = process.argv[2] || "app.js";
installDeps();
wrapApp(entry);

File: instrumentation/python/instrument.py
import os
import subprocess
import sys

def install_opentelemetry_deps():
    subprocess.run([sys.executable, "-m", "pip", "install", "--upgrade", "opentelemetry-distro", "opentelemetry-exporter-otlp"], check=True)

def instrument_python_app(entry_point: str):
    os.environ["OTEL_EXPORTER_OTLP_ENDPOINT"] = "http://otel-collector:4317"
    os.environ["OTEL_TRACES_EXPORTER"] = "otlp"
    os.environ["OTEL_METRICS_EXPORTER"] = "otlp"
    os.environ["OTEL_LOGS_EXPORTER"] = "otlp"
    
    print(f"Running with OpenTelemetry Instrumentation: {entry_point}")
    subprocess.run(["opentelemetry-instrument", "python", entry_point])

if __name__ == "__main__":
    app_entry = sys.argv[1] if len(sys.argv) > 1 else "app.py"
    install_opentelemetry_deps()
    instrument_python_app(app_entry)
